contracts/PermissionLessReserve.sol

```diff
@@ -0,0 +1,596 @@
+
+        return calcRateFromQty(totalSrcAmount, totalDstAmount, getDecimals(src), getDecimals(dest));
+    }
+
+    function trade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        public
+        payable
+        returns(bool)
+    {
+        require((srcToken == ETH_TOKEN_ADDRESS) || (destToken == ETH_TOKEN_ADDRESS));
+        require((srcToken == reserveToken) || (destToken == reserveToken));
+
+        if (validate) {
+            require(conversionRate > 0);
+        }
+
+        uint32 orderId;
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount);
+            orderId = BUY_HEAD_ID;
+        } else {
+            require(srcToken.transferFrom(msg.sender, this, srcAmount));
+            require(msg.value == 0);
+            orderId = SELL_HEAD_ID;
+        }
+
+        uint128 remainingSrcAmount = uint128(srcAmount);
+        uint128 totalDstAmount = 0;
+
+        while (!isNextOrderTail(orderId)) {
+
+            // start with next since head order is dummy order
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount <= remainingSrcAmount) {
+                totalDstAmount += order.dstAmount;
+                remainingSrcAmount -= order.srcAmount;
+                require(takeFullOrder(orderId, srcToken, destToken, order));
+                if (remainingSrcAmount == 0) break;
+            } else {
+                uint128 partialDstQty = order.dstAmount * remainingSrcAmount / order.srcAmount;
+                totalDstAmount += partialDstQty;
+                require(takePartialOrder(orderId, srcToken, destToken, remainingSrcAmount, partialDstQty));
+                remainingSrcAmount = 0;
+                break;
+            }
+        }
+
+        //all orders were successfully taken. send to destAddress
+        if (destToken == ETH_TOKEN_ADDRESS) {
+            destAddress.transfer(totalDstAmount);
+        } else {
+            require(destToken.transfer(destAddress, totalDstAmount));
+        }
+
+        return true;
+    }
+
+    event NewMakeOrder(uint32 orderId, address indexed maker, bool isEthToToken, uint128 srcAmount, uint128 dstAmount);
+
+    function addMakeOrder(address maker, bool isEthToToken, uint128 srcAmount, uint128 dstAmount,
+        uint32 hintPrevOrder) public returns(bool)
+    {
+        require(maker == msg.sender);
+        require(validateOrder(maker, isEthToToken, srcAmount, dstAmount));
+
+        uint32 newID = takeOrderId(maker);
+
+        if (hintPrevOrder != 0) {
+
+            addAfterId(maker, newID, srcAmount, dstAmount, hintPrevOrder, isEthToToken? ETH_TO_TOKEN : TOKEN_TO_ETH);
+        } else {
+
+            if (isEthToToken) {
+                add(maker, newID, srcAmount, dstAmount, BUY_HEAD_ID, ETH_TO_TOKEN);
```

Comment:
seems you would only need BUY/SELL or ETH_TO_TOKEN/TOKEN_TO_ETH, why do you need both?